providers/client_socket.js

```diff
@@ -148,7 +148,10 @@ ClientSocket.prototype.close = function() {
   // Delete transport so getInfo doesn't think we are connected
   delete this.transport;
   if (typeof this.onDisconnect === 'function') {
-    this.onDisconnect();
+    this.onDisconnect(continuation);
+    delete this.onDisconnect;
+  } else if (typeof continuation === 'function') {
+    continuation();
```

Comment:
If onDisconnect is defined the continuation gets passed through and should be called by it, at least as it's currently structured.